### ISSUE TITLE
Performance issues and false positives in `v4.0.0-alpha`

### ISSUE THREAD

**Issue by robert-barahona** [francoismassart/eslint-plugin-tailwindcss#448]
After spending some time on eslint-plugin-tailwindcss [v4.0.0-alpha.2](https://github.com/francoismassart/eslint-plugin-tailwindcss/releases/tag/v4.0.0-alpha.2) on my monorepo

I found some relevant issues on my end:

1. **Performance.** The lint takes 10+ minutes to finish in the bigger package of my monorepo.
2. **False positives on tailwind-variants.** It walks every string inside tv() calls as a classname, so variant identifiers (primary, pink, md, true, false, etc.) in defaultVariants/compoundVariants get flagged as invalid Tailwind classes. No semantic awareness of tv()'s shape. (see attached ss)

<img width="1174" height="113" alt="Screenshot 2026-05-06 at 09 36 21" src="https://github.com/user-attachments/assets/a422426f-dc9a-48ff-bf17-60a9c8ecd2ac" />

3. **API Changes.** "no-custom-classname" now uses a whitelist option (regex strings of classnames to ignore) instead of the previous cssFiles (globs of CSS files to scan) and the plugin does not detect our custom CSS classes, so there are a lot of false positives. (see attached ss)

<img width="1371" height="95" alt="Screenshot 2026-05-06 at 10 19 18" src="https://github.com/user-attachments/assets/13b94795-23ff-4f07-9ee6-6a5b8236d0cb" />

---

This is my config:

```ts
import { dirname, resolve } from "path"
import { fileURLToPath } from "url"

import tailwindcss from "eslint-plugin-tailwindcss"

const __dirname = dirname(fileURLToPath(import.meta.url))


const globalConfig = [
  {
    files: ["**/*.ts", "**/*.tsx"],

    plugins: {
      tailwindcss: tailwindcss
    },

    settings: {
      tailwindcss: {
        callees: ["classNames", "cn", "tv"],
        cssConfigPath: resolve(__dirname, "styles/tailwind.css")
      }
    },

    rules: {
      "tailwindcss/no-custom-classname": "error",
    }
  },
]
```

**Comment by francoismassart** [comment]
Hi @robert-barahona,
Thank you for your feedback.

I'm back from 2 weeks in amazing Japan.

Here are some feedbacks:

**1. Performances**

Could you run each rule one by one while using timing options such as:

`npm run script-name-rule-classnames-order --timing`
`npm run script-name-rule-no-contradicting-classname --timing`
`npm run script-name-rule-no-custom-classname --timing`

and/or

`time npm run script-name-rule-classnames-order`
`time npm run script-name-rule-no-contradicting-classname`
`time npm run script-name-rule-no-custom-classname`

**2. False positives on tailwind-variants**

I commited a [fix for the false positive issue](https://github.com/francoismassart/eslint-plugin-tailwindcss/pull/447/changes/7de198673a8af9ca83aa491b1709f30920dd48ae) you are experiencing.

I'll let you know once it is available in pre-release.

**3. API Changes**

For now, I did not plan to include `cssFiles` option but it could be done in the future.

**Comment by aaronabf** [comment]
> I'm back from 2 weeks in amazing Japan.

Same. It was so nice.

Thanks for the assistance here!

**Comment by robert-barahona** [comment]
Hey @francoismassart thanks for your reply

**1. Performance** 

Here are the results.


classnames-order:
```
✖ 91 problems (91 errors, 0 warnings)
yarn lint:check  251.79s user 34.00s system 136% cpu 3:29.59 total
```

no-custom-classname:
```
✖ 89 problems (89 errors, 0 warnings)
yarn lint:check  919.52s user 104.47s system 134% cpu 12:42.77 total
```

no-contradicting-classname:
```
✖ 6 problems (6 errors, 0 warnings)
yarn lint:check  880.34s user 99.24s system 134% cpu 12:08.61 total
```

**2. False positives on tailwind-variants**

> I commited a [fix for the false positive issue](https://github.com/francoismassart/eslint-plugin-tailwindcss/pull/447/changes/7de198673a8af9ca83aa491b1709f30920dd48ae) you are experiencing.
>  I'll let you know once it is available in pre-release.

Great, thank you!

**3. API Changes**

> For now, I did not plan to include cssFiles option but it could be done in the future.

I see, but shouldn't the CSS files imported into the main input file be detected automatically?

```js
// eslint.config.mjs
...
settings: {
  tailwindcss: {
    callees: ["classNames", "cn", "tv"],
    cssConfigPath: resolve(__dirname, "main.css")
  }
},
...
```

```css
/* main.css */

@import "./landing.css";
```

```css
/* landing.css */

.landing-btn {
  ...
}
```

Expected result: `landing-btn` shouldn't be reported in `tailwindcss/no-custom-classname`

**Comment by f1yingbanana** [comment]
Hate to nag, but any update on this? I have to disable this in my editor if I want to run eslint on save and it makes things really slow in CI workflows :/

**Comment by francoismassart** [comment]
Hi @f1yingbanana,
I've been busy migrating the existing rules but I did not yet work on the perfs.

Make sure to check https://github.com/francoismassart/eslint-plugin-tailwindcss/releases to know when a new alpha is ready for testing

**Comment by francoismassart** [comment]
Hi @robert-barahona,
**I just released a new alpha which should enhance the `classnames-order` rule.**

Could you:

1.  `npm i eslint-plugin-tailwindcss@4.0.0-alpha.4`
2. Run the TIMING lint for the `classnames-order` rule
3. Share the results

Last time, you had `classnames-order` rule:

```
✖ 91 problems (91 errors, 0 warnings)
yarn lint:check  251.79s user 34.00s system 136% cpu 3:29.59 total
```

Also, I would like to know more about your set up:

- IDE
- Machine you are using (please also include CPU + RAM)
- Are you using a Monorepo ?
- With NX or a similar setup ?
- Some extract of your codebase...
  - Vue ? React ? Angular ?
  - How do you include your classnames ? (simple attributes, helper like `cva`, etc.)
- How large is your repo:
  - In terms of parsed files
  - Number of classnames attributes
- Example of you Tailwind CSS config
  - In need to know how complex it is
  - How it is structured...
  - Do you reset it all ?

Even better would be to set up a repo with a demo codebase that reflects your project in order to let me play with it and improve the perfs.

Thank you for your help.

**Comment by francoismassart** [comment]
I've finished the first enhancements on all existing rules: 
https://www.npmjs.com/package/eslint-plugin-tailwindcss/v/4.0.0-alpha.5

Please try it out and share your feedback!

`npm i eslint-plugin-tailwindcss@4.0.0-alpha.5`

**Comment by f1yingbanana** [comment]
Can confirm alpha 5 fixed the issue! Thanks! :)

Just curious, what was the bottleneck btw?

```
tailwindcss/no-custom-classname         |   970.593 |     5.3%
tailwindcss/enforces-shorthand          |   391.335 |     2.2%
```

**Comment by robert-barahona** [comment]
@francoismassart Great! I'll take a look this Monday and will let you know

**Comment by robert-barahona** [comment]
Hey @francoismassart! 

I tried with `4.0.0-alpha.5` and there's a huge difference in performance, here is the comparison:

### classnames-order:

Before:

```
✖ 91 problems (91 errors, 0 warnings)
yarn lint:check  251.79s user 34.00s system 136% cpu 3:29.59 total
```

After:

```
✖ 35 problems (35 errors, 0 warnings)
yarn lint:check  44.34s user 18.82s system 108% cpu 58.459 total
```

### no-custom-classname:

Before:

```
✖ 89 problems (89 errors, 0 warnings)
yarn lint:check  919.52s user 104.47s system 134% cpu 12:42.77 total
```

After:

```
✖ 7 problems (2 errors, 5 warnings)
yarn lint:check  293.29s user 119.05s system 358% cpu 1:55.14 total
```

### no-contradicting-classname:

Before:

```
✖ 6 problems (6 errors, 0 warnings)
yarn lint:check  880.34s user 99.24s system 134% cpu 12:08.61 total
```

After:

```
✖ 4 problems (4 errors, 0 warnings)
yarn lint:check  85.02s user 40.92s system 180% cpu 1:09.90 total
```

---

And I don't get any more false positives with `"no-custom-classname"` 🥳 

Feel free to close this issue, thank you!

**Comment by francoismassart** [comment]
v4 is available
